// acies/lib/paths.js
"use strict";
exports.__esModule = true;
var Path = require("./path");
function boundingBox(p) {
    var box = Path.boundingBox(p[0]);
    for (var _i = 0, p_1 = p; _i < p_1.length; _i++) {
        var path = p_1[_i];
        box = box.extend(Path.boundingBox(path));
    }
    return box;
}
exports.boundingBox = boundingBox;
function transform(p, matrix) {
    var result = [];
    for (var _i = 0, p_2 = p; _i < p_2.length; _i++) {
        var path = p_2[_i];
        result.push(Path.transform(path, matrix));
    }
    return result;
}
exports.transform = transform;
function chop(p, step) {
    var result = [];
    for (var _i = 0, p_3 = p; _i < p_3.length; _i++) {
        var path = p_3[_i];
        result.push(Path.chop(path, step));
    }
    return result;
}
exports.chop = chop;
function filterPaths(p, f) {
    var result = [];
    for (var _i = 0, p_4 = p; _i < p_4.length; _i++) {
        var path = p_4[_i];
        result.push.apply(result, Path.filterPaths(path, f));
    }
    return result;
}
exports.filterPaths = filterPaths;
function simplify(p, threshold) {
    var result = [];
    for (var _i = 0, p_5 = p; _i < p_5.length; _i++) {
        var path = p_5[_i];
        result.push(Path.simplify(path, threshold));
    }
    return result;
}
exports.simplify = simplify;
function toSVG(p, width, height) {
    var lines = [];
    lines.push("<svg width=\"" + width + "\" height=\"" + height + "\" version=\"1.1\" baseProfile=\"full\" xmlns=\"http://www.w3.org/2000/svg\">");
    lines.push("<g transform=\"translate(0," + height + ") scale(1,-1)\">");
    for (var _i = 0, p_6 = p; _i < p_6.length; _i++) {
        var path = p_6[_i];
        lines.push(Path.toSVG(path));
    }
    lines.push("</g></svg>");
    return lines.join("\n");
}
exports.toSVG = toSVG;
